gui/widgets/heatmap_widget.py

```diff
@@ -0,0 +1,182 @@
+
+
+class HeatmapWidget(MplCanvas):
+    def __init__(self, parent: "ControlMain", width=5, height=4, dpi=100, data=None, cell_results=None):
+        super().__init__(parent, width, height, dpi)
+        self.mpl_connect("button_press_event", self.on_click)
+        self.mpl_connect("motion_notify_event", self.on_hover)
+        self._parent = parent
+        self.data = data
+        self.cell_results = cell_results
+        self.highlighted_patches = {}
+        self.render_heatmap()
+
+    def render_heatmap(self):
+        if self.data is None:
+            return
+
+        # Create the heatmap
+        cax = self.axes.imshow(self.data, cmap="inferno", origin="upper")
+
+        y_ticks = np.arange(self.data.shape[0])
+        x_ticks = np.arange(self.data.shape[1])
+
+        # Label offset: adjust ticks to be between points
+        self.axes.set_xticks(x_ticks)
+        self.axes.set_yticks(y_ticks)
+
+        self.axes.set_xlim([x_ticks[0] - 0.5, x_ticks[-1] + 0.5])
+        self.axes.set_ylim([y_ticks[-1] + 0.5, y_ticks[0] - 0.5])
+
+        divider = make_axes_locatable(self.axes)
+        cax_cb = divider.append_axes(
+            "right", size="5%", pad=0.05
+        )  # Adjust size and padding as needed
+        self.colorbar = self.figure.colorbar(cax, cax=cax_cb)
+
+        # Create a text annotation for the tooltip, initially hidden
+        self.tooltip = self.axes.text(
+            -2.5,
+            0.95,
+            "",
+            color="white",
+            backgroundcolor="black",
+            ha="center",
+            va="center",
+            fontsize=10,
+            bbox=dict(facecolor="black", alpha=0.8),
+        )
+        self.tooltip.set_visible(False)
+
+        # Create a rectangle for highlighting cells, initially hidden
+        self.highlight = Rectangle(
+            (0, 0), 1, 1, linewidth=2, edgecolor="white", facecolor="none"
+        )
+        self.axes.add_patch(self.highlight)
+        self.highlight.set_visible(False)
+
+        # Redraw the canvas
+        self.draw()
+        self.figure.tight_layout()
+
+    def on_click(self, event):
+        if self.data is None or event.inaxes is None:
+            return
+        x, y = int(round(event.xdata)), int(round(event.ydata))
+        col, row = int(np.floor(x)), int(np.floor(y))
+        print(f"Clicked at: ({event.xdata}, {event.ydata})")
+        if 0 <= row < self.data.shape[0] and 0 <= col < self.data.shape[1]:
+            value = self.data[row, col]
+            if event.dblclick:
+                self.highlight_patch(row, col)
+            else:
+                self.show_diffraction_image(row, col)
+
+    def show_diffraction_image(self, row, col):
+        if not self.cell_results:
+            return
+        print(row, col, self.data.shape[0], self.data.shape[1])
+        flattened_index = calculate_flattened_index(row, col, self.data.shape[0], self.data.shape[1])
+        print(flattened_index)
+        print(len(self.cell_results))
```

Comment:
do you want these prints in the production code?

Reply:
Removed print statement in https://github.com/NSLS2/lsdc/pull/414/commits/df88266519a813a998fc82e43db3f3976b726f23